src/telemetry_formatter.rs

```diff
@@ -0,0 +1,218 @@
+            os: System::name().unwrap_or("unknown".to_string()),
+            os_version: System::kernel_version().unwrap_or("unknown".to_string()),
+            trace_id: Uuid::new_v4().to_string(),
+            triple: format!(
+                "{}-{}-{}",
+                match std::env::consts::ARCH {
+                    arch @ ("aarch64" | "x86_64") => arch,
+                    _ => "unknown",
+                },
+                match std::env::consts::OS {
+                    "macos" => "apple",
+                    _ => "unknown",
+                },
+                match std::env::consts::OS {
+                    "macos" => "darwin",
+                    "linux" => "linux-gnu",
+                    _ => "unknown",
+                }
+            ),
+        }
+    }
+}
+
+impl<S, N> FormatEvent<S, N> for TelemetryFormatter
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+    N: for<'a> FormatFields<'a> + 'static,
+{
+    /// Formats an `Event`.
+    ///
+    /// This function formats an `Event` into a `tracing` `Writer` with the following format:
+    ///
+    /// <timestamp> <level> <triple>:<os>:<os-version> <crate>:<version>:<file> <trace-id> <spans> <fields>
+    ///
+    /// e.g:
+    ///
+    /// ```text
+    /// 2021-08-31T14:00:00.000000Z ERROR x86_64-apple-darwin:Arch Linux:6.12.3-arch1-1 fuel-telemetry:0.1.0:fuelup/src/main.rs ddfc7485-c40f-4e3f-8203-704cccbd7475 root:span1:span2: field1="val1" "A test message"
+    /// ```
+    fn format_event(
+        &self,
+        ctx: &FmtContext<'_, S, N>,
+        mut writer: Writer<'_>,
+        event: &Event<'_>,
+    ) -> std::fmt::Result {
+        let mut wants_telemetry = false;
+
+        // Check if the event wants telemetry enabled
+        //
+        // Going from leaf span to the root, we check if any of the spans have a
+        // `telemetry` field set to `false`. If so, short-circuit return.
+        //
+        // If a `telemetry` field is set to `true`, we set the `wants_telemetry`
+        // flag to `true` and then break.
+        for span in ctx.event_scope().into_iter().flatten() {
+            if let Some(fields) = span.extensions().get::<FormattedFields<N>>() {
+                // A regex to parse the `telemetry` field from the span's fields
+                static TELEMETRY_SETTING_REGEX: LazyLock<Result<Regex>> =
+                    LazyLock::new(|| Ok(regex::Regex::new(r"telemetry\s*=\s*(true|false)")?));
+
+                let telemetry_setting = TELEMETRY_SETTING_REGEX
+                    .as_ref()
+                    .ok()
+                    .and_then(|regex| regex.captures(fields.fields.as_str()))
+                    .and_then(|caps| caps.get(1))
+                    .map(|m| m.as_str());
+
+                if let Some("false") = telemetry_setting {
+                    return Ok(());
+                } else if let Some("true") = telemetry_setting {
+                    wants_telemetry = true;
+                    break;
+                }
```

Comment:
nit: use `match`; also what if this is not `true` or `false`?

Reply:
Nice. Yep, it does look nicer, thanks! aaf7aa587bf16